subgraphs/minimal-legacy-exchange/src/pricing.ts

```diff
@@ -0,0 +1,137 @@
+import { Pool, Token, TokenKpi } from '../generated/schema'
+import { Factory as FactoryContract } from '../generated/templates/Pair/Factory'
+import { getPoolKpi } from './functions/pool-data'
+
+export const factoryContract = FactoryContract.bind(FACTORY_ADDRESS)
+
+export function getNativePriceInUSD(): BigDecimal {
+  let count = 0
+  let weightdPrice = BigDecimal.fromString('0')
+  let nativeReserve = BigDecimal.fromString('0')
+  let stablePrices: BigDecimal[] = []
+  let nativeReserves: BigDecimal[] = []
+
+  for (let i = 0; i < STABLE_POOL_ADDRESSES.length; i++) {
+    const address = STABLE_POOL_ADDRESSES[i]
+
+    const stablePool = Pool.load(address)
+    if (stablePool === null) {
+      continue
+    }
+    const stablePoolKpi = getPoolKpi(address)
+
+    if (
+      (stablePool.token0 == NATIVE_ADDRESS && stablePoolKpi.reserve0.lt(MINIMUM_NATIVE_LIQUIDITY)) ||
+      (stablePool.token1 == NATIVE_ADDRESS && stablePoolKpi.reserve1.lt(MINIMUM_NATIVE_LIQUIDITY))
+    ) {
+      continue
+    }
+
+    const stableFirst = STABLE_TOKEN_ADDRESSES.includes(stablePool.token0)
+
+    nativeReserve = nativeReserve.plus(!stableFirst ? stablePoolKpi.reserve0 : stablePoolKpi.reserve1)
+
+    nativeReserves.push(!stableFirst ? stablePoolKpi.reserve0 : stablePoolKpi.reserve1)
+
+    stablePrices.push(stableFirst ? stablePoolKpi.token0Price : stablePoolKpi.token1Price)
+
+    count = count + 1
+  }
+
+  if (count > 0) {
+    for (let j = 0; j < count; j++) {
+      const price = stablePrices[j]
+      const weight = nativeReserves[j].div(nativeReserve)
+      weightdPrice = weightdPrice.plus(price.times(weight))
+    }
+  }
+
+  return weightdPrice
+}
+
+/**
+ * Updates the token KPI price for the given token.
+ * Find the pool that contains the most liquidity and is safe from circular price dependency,
+ * (e.g. if DAI is priced off USDC, then USDC cannot be priced off DAI)
+ * @param tokenAddress The address of the token kpi to update
+ * @returns
+ */
+export function updateTokenKpiPrice(tokenAddress: string): TokenKpi {
+  const token = getOrCreateToken(tokenAddress)
+  const currentTokenKpi = getTokenKpi(tokenAddress)
+  if (token.id == NATIVE_ADDRESS) {
+    if (!currentTokenKpi.derivedETH.equals(BIG_DECIMAL_ONE)) {
+      currentTokenKpi.derivedETH = BIG_DECIMAL_ONE
+      currentTokenKpi.save()
+    }
+    return currentTokenKpi
+  }
+
+  const pools = currentTokenKpi.pools
+
+  let pricedOffToken = ''
+  let mostReseveEth = BIG_DECIMAL_ZERO
+  let currentPrice = BIG_DECIMAL_ZERO
+
+  for (let i = 0; i < pools.length; ++i) {
+    const poolAddress = pools[i]
+    const pool = Pool.load(poolAddress)
+    if (pool === null) {
+      continue // Not created yet
+    }
+    const poolKpi = getPoolKpi(poolAddress)
+    const poolToken0Kpi = getTokenKpi(pool.token0)
+    const poolToken1Kpi = getTokenKpi(pool.token1)
+
+    if (
+      pool.token0 == token.id &&
+      poolToken1Kpi.pricedOffToken != token.id &&
```

Comment:
on L100 and L114, the `priceOffToken` field is included in in the criteria to ignore the current pool from pricing if it's the same as the current token we are trying to price.